### PR TITLE
 [SR GeekCamp] Select the smallest data type for four arithmetic operations

### PR DESCRIPTION
**Pull request by mchades**
finished issue #4038

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 53 / 80 (66.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|16|38|42.11%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|30|35|85.71%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|6|6|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 1]
```diff
@@ -1354,6 +1357,79 @@ public Type getNumResultType() {
         }
     }
 
+    /**
+     * return a largest order of magnitude between this and t2
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:1361]: It looks like return the bigger type, not the order of magnitude

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 2]
```diff
@@ -1354,6 +1357,79 @@ public Type getNumResultType() {
...
+     * return Type.INVALID if none of them are numeric
+     *
+     * @param t2 Type to compared with
+     * @return a largest order of magnitude between this and t2
+     */
+    public Type max(Type t2) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:1367]: suggest use static method like getCommonType

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 3]
```diff
@@ -389,6 +420,55 @@ public static ConstantOperator divideDecimal(ConstantOperator first, ConstantOpe
         return createDecimalConstant(first.getDecimal().divide(second.getDecimal()));
     }
 
+    @FEFunction(name = "int_divide", argTypes = {"TINYINT", "TINYINT"}, returnType = "TINYINT")
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:423]: Plz add test to cover this functions

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 4]
```diff
@@ -1354,6 +1357,79 @@ public Type getNumResultType() {
...
+        } else {
+            return Type.INVALID;
+        }
+    }
+
+    /**
+     * Raise by an order of magnitude
+     * return Type.INVALID if it is not numeric
+     */
+    public Type up() {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:1396]: suggest use static method like public static Type getBiggerType(Type t)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java (hunk 3 above)
Comment (mchades) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:423]: OK. Already move those functions add in another PR #4807

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 39 / 44 (88.64%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/ArithmeticExpr.java|29|33|87.88%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|8|87.50%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java [hunk 5]
```diff
@@ -425,6 +425,60 @@ public static Type findCommonType(Type t1, Type t2) {
...
+            return t1;
+        } else if (PrimitiveType.TINYINT.ordinal() <= pt2.ordinal() &&
+                pt2.ordinal() <= PrimitiveType.INT.ordinal()) {
+            return t2;
+        }else {
+            return Type.INVALID;
+        }
+    }
+
+    public  static Type getBiggerType(Type t) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java:458]:
```suggestion
    public static Type getBiggerType(Type t) {
```

File: fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java [hunk 6]
```diff
@@ -425,6 +425,60 @@ public static Type findCommonType(Type t1, Type t2) {
...
+                (PrimitiveType.TINYINT.ordinal() <= pt2.ordinal() &&
+                        pt2.ordinal() <= PrimitiveType.INT.ordinal())) {
+            return (pt1.ordinal() > pt2.ordinal()) ? t1 : t2;
+        } else if (PrimitiveType.TINYINT.ordinal() <= pt1.ordinal() &&
+                pt1.ordinal() <= PrimitiveType.INT.ordinal()) {
+            return t1;
+        } else if (PrimitiveType.TINYINT.ordinal() <= pt2.ordinal() &&
+                pt2.ordinal() <= PrimitiveType.INT.ordinal()) {
+            return t2;
+        }else {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java:453]:
```suggestion
        } else {
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 39 / 44 (88.64%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/ArithmeticExpr.java|29|33|87.88%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|8|87.50%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java [hunk 7]
```diff
@@ -425,6 +425,60 @@ public static Type findCommonType(Type t1, Type t2) {
         }
     }
 
+    public static Type getCommonType(Type t1, Type t2) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java:428]: Can be replace all `ArithmeticExpr.findCommonType`?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 39 / 44 (88.64%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/ArithmeticExpr.java|29|33|87.88%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|7|8|87.50%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java (hunk 7 above)
Comment (mchades) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java:428]: I don't know what the side effects will be, I think it should be replaced in another pr.

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java:428]: > Can be replace all `ArithmeticExpr.findCommonType`?

@Seaven This function is still used by old planner, we can delete it later

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ArithmeticExpr.java:428]: > > Can be replace all `ArithmeticExpr.findCommonType`?
> 
> @Seaven This function is still used by old planner, we can delete it later

OK, record issue to remove it later plz

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**